chore(individual-pkg-gen): remove L1s from alpha modules

madeline-k opened the pull request:
The PR contains the following changes:
1. Skip copying the generated L1 files into the alpha modules
2. Re-write imports in the alpha modules that reference L1s to reference aws-cdk-lib/aws-<service>
  These imports come in the following formats: `./<service>.generated`, `../<service>.generated`, `../lib/<service>.generated`, `./<service>-canned-metrics.generated`. All of these formats get converted to `aws-cdk-lib/aws-<service>`
3. Don't export generated L1s in the `index.ts` files of the alpha modules.

Closes #15587 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 5c2890036ef6f07bf89c8f4d95f69e6ebfb2a13e
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bf0e6a9f-cedf-4dc3-8afa-17947d1b0468%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch commented on packages/aws-cdk-migration/lib/rewrite.ts:
```diff
@@ -108,6 +113,12 @@ function updatedLocationOf(modulePath: string, options: RewriteOptions): string
   }
 
+  if (options.isInAlphaPackage && modulePath.endsWith('generated')) {
+    const modulePathSplit = modulePath.split(/[.\/]/);
```
The forward-slash doesn't need to be escaped here.

```suggestion
    const modulePathSplit = modulePath.split(/[./]/);
```

njlynch commented on packages/aws-cdk-migration/lib/rewrite.ts:
```diff
@@ -108,6 +113,12 @@ function updatedLocationOf(modulePath: string, options: RewriteOptions): string
+  if (options.isInAlphaPackage && modulePath.endsWith('generated')) {
+    const modulePathSplit = modulePath.split(/[.\/]/);
+    // The following line takes the 2nd to last item in modulePathSplit, which will always be the basename of the module e.g. api-gatewayv2.
+    return `aws-cdk-lib/aws-${modulePathSplit[modulePathSplit.length-2].replace('-canned-metrics', '')}`;
```
Ooh. This raises a fantastic question we haven't addressed (read: I hadn't thought about) yet.

`cfn2ts` generates three types of files: the L1s, "Augmentations" (for metrics), and Canned Metrics:
https://github.com/aws/aws-cdk/blob/2023004cc941a0e7a908bf3c90ad9887c6679564/tools/cfn2ts/lib/index.ts#L21-L30

All three of these have `generated` in their file names. The L1s absolutely need to be included in `aws-cdk-lib` and removed from the alpha modules. The "Augmentations" directly reference `IFoo` and `FooBase`, and so would need to be included in the alpha modules (and removed from `aws-cdk-lib`). The Canned Metrics are a little more ambiguous: the define static factory methods that produce inputs for a `cloudwatch.Metric`. This could be theoretically useful for the L1s as well, but are really more of an L2 thing and I would suggest probably belong in the alpha modules.

My opinion would be to only remove (and rewrite imports) for the L1s, and leave the two generated metrics files as-is. That also implies we need a new task to track removing the generated metrics files from `aws-cdk-lib` for alpha modules. _sigh_

madeline-k replied:
Created a new task for this https://github.com/aws/aws-cdk/issues/16369

madeline-k replied:
In this implementation, the canned metrics and augmentations are being included in the alpha modules now.

njlynch reviewed (changes requested):
It's unfortunate the `individual-pkg-gen` package doesn't currently have tests, but can you add a test for the `rewrite` changes in `aws-cdk-migration`?

nija-at commented on packages/aws-cdk-migration/lib/rewrite.ts:
```diff
@@ -8,6 +8,11 @@ export interface RewriteOptions {
   readonly customModules?: { [moduleName: string]: string };
+
+  /**
+   * Optional flag to set for rewriting imports in alpha packages. When true, this will rewrite imports of generated L1s to reference aws-cdk-lib.
```
Would be nice to provide a tiny example in the tsdoc.

nija-at commented on packages/aws-cdk-migration/lib/rewrite.ts:
```diff
@@ -8,6 +8,11 @@ export interface RewriteOptions {
+  /**
+   * Optional flag to set for rewriting imports in alpha packages. When true, this will rewrite imports of generated L1s to reference aws-cdk-lib.
+   */
+  readonly isInAlphaPackage?: boolean;
```
Name feels leaky. Not sure `aws-cdk-migration` should be aware of alpha packages.
A more direct name for what this feature does would be better.

Something like `useCfnConstructs` or `useBaseCfnConstructs`?

madeline-k replied:
I opted for `rewriteCfnImports` to keep it as specific as possible. Let me know what you think

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: b14b4afe5674e16fc49574f2bb913b0588353b94
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=18cc7469-0a94-4ebe-8d4b-801069d986b6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k commented on tools/individual-pkg-gen/copy-files-removing-deps.ts:
```diff
@@ -73,10 +77,22 @@ function transformPackages(): void {
+            continue;
+          }
+          resultFileLines.push(line);
+        }
```
Done

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 178b8ef34bce1351708fcaaed40f79936efed182
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1bf7fcbe-4c3a-42ed-b253-230b2d61b368%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k commented on packages/aws-cdk-migration/lib/rewrite.ts:
```diff
@@ -108,6 +113,12 @@ function updatedLocationOf(modulePath: string, options: RewriteOptions): string
+  if (options.isInAlphaPackage && modulePath.endsWith('generated')) {
+    const modulePathSplit = modulePath.split(/[.\/]/);
+    // The following line takes the 2nd to last item in modulePathSplit, which will always be the basename of the module e.g. api-gatewayv2.
+    return `aws-cdk-lib/aws-${modulePathSplit[modulePathSplit.length-2].replace('-canned-metrics', '')}`;
```
I've removed the `.replace('-canned-metrics','')` part now based on Nick's suggestion. I agree with you this was hacky, and it still is a little bit. Open to any other suggestions you have to improve it!

madeline-k commented:
I am running into an issue now with imports like this:

https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-redshift/test/cluster.test.ts#L6

```
import { CfnCluster, Cluster, ClusterParameterGroup, ClusterSubnetGroup, ClusterType } from '../lib';
```

This one is currently failing the PR Build, and there are probably more. 

I see two solutions here:
1. enforce in v1 that Cfn imports are separate, and actually directly reference the generated file, not just lib. That would rewrite the above to the following on master, and have a linter rule that requires Cfn imports to be separate. With this option, the logic I've written here won't have to change.
```
import { Cluster, ClusterParameterGroup, ClusterSubnetGroup, ClusterType } from '../lib';
import { CfnCluster } from '../lib/redshift.generated';
```

2. In `rewrite.ts` and `copy-files-removing-deps.ts`, write logic to directly transform to the below when running `transform.sh`.
```
import { Cluster, ClusterParameterGroup, ClusterSubnetGroup, ClusterType } from '../lib';
import { CfnCluster } from 'aws-cdk-lib/aws-redshift';
```

I think option 2 would be ideal, but I am concerned that changing one line into two in the v2 rewrite tools will get messy.

njlynch commented:
> 1. enforce in v1 that Cfn imports are separate, and actually directly reference the generated file, not just lib. That would rewrite the above to the following on master, and have a linter rule that requires Cfn imports to be separate. With this option, the logic I've written here won't have to change.

It appears there are 11 usages of loading `Cfn*` types from something that's not `@aws-cdk/lib` or `*.generated`:

```sh
% git grep "import .*\bCfn.*" | grep -v generated | grep -v '@aws-cdk/core'
packages/@aws-cdk/aws-appmesh/lib/header-match.ts:import { CfnRoute } from './index';
packages/@aws-cdk/aws-athena/test/athena.test.ts:import { CfnWorkGroup } from '../lib';
packages/@aws-cdk/aws-athena/test/integ.workgroup.ts:import { CfnWorkGroup } from '../lib';
packages/@aws-cdk/aws-cloudfront/test/test-origin.ts:import { CfnDistribution, IOrigin, OriginBase, OriginBindConfig, OriginBindOptions, OriginProps, OriginProtocolPolicy } from '../lib';
packages/@aws-cdk/aws-iam/test/policy.test.ts:import { AnyPrincipal, CfnPolicy, Group, Policy, PolicyDocument, PolicyStatement, Role, ServicePrincipal, User } from '../lib';
packages/@aws-cdk/aws-msk/lib/cluster.ts:import { CfnCluster, KafkaVersion } from './';
packages/@aws-cdk/aws-redshift/test/cluster.test.ts:import { CfnCluster, Cluster, ClusterParameterGroup, ClusterSubnetGroup, ClusterType } from '../lib';
packages/@aws-cdk/aws-sam/test/application.test.ts:import { CfnApplication } from '../lib';
packages/@aws-cdk/cfnspec/lib/index.ts:import { CfnLintFileSchema } from './_private_schema/cfn-lint';
packages/awslint/lib/rules/module.ts:import { CfnResourceReflection } from './cfn-resource';
packages/awslint/lib/rules/resource.ts:import { CfnResourceReflection } from './cfn-resource';
```

Of those, 7 are only importing the L1s, so there's only 4 statements that would conflict with the proposed rule (as long as we exclude imports from `@aws-cdk/core`).

*However*, I worry a bit about us dodging this problem for named imports and missing the complexity of module imports:

```ts
import * as foo from '@aws-cdk/aws-foo';
/// ...
new foo.Flubber(...);
new foo.CfnFizzBuzz(...);
///
```

The above case will _*also*_ require us to create the new import statements. If we're going to have to do it for that case, we might just want to handle it for both cases.

So Option 2 is probably preferable in the long run.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 1c039762350fa46ccbd9303f0872a07d078c090d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f2d480a7-32b1-453b-b138-5c91dbfa6ea5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

madeline-k commented:
After a lot of investigation into the options here, I think the best thing is to require that the L1 imports that we support rewriting in this tool are in a "normal" format. The three options are:

1. enforce in v1 that Cfn imports are separate, and actually directly reference the generated file, not just lib. That would rewrite the above to the following on master, and have a linter rule that requires Cfn imports to be separate. With this option, the logic I've written here won't have to change. Implemented in this PR: https://github.com/aws/aws-cdk/pull/16481 

2. In rewrite.ts and copy-files-removing-deps.ts, write logic to directly transform Cfn imports using text manipulation. I implemented this one in another PR: https://github.com/aws/aws-cdk/pull/16482. It is super hacky, and I don't really trust it to not cause headaches for us in the future. Open to suggestions on how to make this one better. 

3. In rewrite.ts and copy-files-removing-deps.ts, write logic to directly transform the Cfn imports (and all imports that need updating) using `ts.updateXX()` apis to manipulate the AST nodes. I explored this option in this PR: https://github.com/aws/aws-cdk/pull/16483 It is not really working fully yet. But I have hit a roadblock, more description in the PR. 

I think we should go with option 1. @njlynch, @rix0rrr, @nija-at, thoughts?

Closing this PR in favor of one of the above 3 :)

nija-at commented:
Option 1 seems the best to me. Trying to manipulate typescript text or nodes is just going to make this harder. Option 3 if it is simple and clean. Option 2 just sounds bad.
